File: lca/src/main/java/com/dreampany/lca/data/model/Exchange.java
package com.dreampany.lca.data.model;

import android.arch.persistence.room.Entity;
import android.arch.persistence.room.Ignore;
import android.arch.persistence.room.Index;
import android.os.Parcel;
import android.support.annotation.NonNull;

import com.dreampany.frame.data.model.Base;
import com.google.common.base.Objects;

/**
 * Created by Hawladar Roman on 24/6/18.
 * Dreampany Ltd
 * dreampanymail@gmail.com
 */
@Entity(indices = {@Index(value = {"id"}, unique = true)},
        primaryKeys = {"id"})
public class Exchange extends Base {

    private String exchange;
    private String fromSymbol;
    private String toSymbol;
    private double volume24h;
    private double volume24hTo;

    @Ignore
    public Exchange() {

    }

    public Exchange(long id) {
        this.id = id;
    }

/*    public Exchange(@NonNull String exchange, @NonNull String fromSymbol, @NonNull String toSymbol) {
        this.exchange = exchange;
        this.fromSymbol = fromSymbol;
        this.toSymbol = toSymbol;
    }*/

    @Ignore
    private Exchange(Parcel in) {
        super(in);
        exchange = in.readString();
        fromSymbol = in.readString();
        toSymbol = in.readString();
        volume24h = in.readDouble();
        volume24hTo = in.readDouble();
    }

    @Override
    public void writeToParcel(@NonNull Parcel dest, int flags) {
        super.writeToParcel(dest, flags);
        dest.writeLong(id);
        dest.writeString(exchange);
        dest.writeString(fromSymbol);
        dest.writeString(toSymbol);
        dest.writeDouble(volume24h);
        dest.writeDouble(volume24hTo);
    }

    public static final Creator<Exchange> CREATOR = new Creator<Exchange>() {
        @Override
        public Exchange createFromParcel(Parcel in) {
            return new Exchange(in);
        }

        @Override
        public Exchange[] newArray(int size) {
            return new Exchange[size];
        }
    };

    @Override
    public String toString() {
        return "Exchange{" +
                "exchange=" + exchange +
                ", toSymbol='" + toSymbol + '\'' +
                ", fromSymbol='" + fromSymbol + '\'' +
                '}';
    }

    public void setExchange(String exchange) {
        this.exchange = exchange;
    }

    public void setFromSymbol( String fromSymbol) {
        this.fromSymbol = fromSymbol;
    }

    public void setToSymbol( String toSymbol) {
        this.toSymbol = toSymbol;
    }

    public void setVolume24h(double volume24h) {
        this.volume24h = volume24h;
    }

    public void setVolume24hTo(double volume24hTo) {
        this.volume24hTo = volume24hTo;
    }

    public String getExchange() {
        return exchange;
    }

    public String getFromSymbol() {
        return fromSymbol;
    }

    public String getToSymbol() {
        return toSymbol;
    }

    public double getVolume24h() {
        return volume24h;
    }

    public double getVolume24hTo() {
        return volume24hTo;
    }
}

File: crypto/src/main/java/com/dreampany/crypto/data/model/Exchange.kt
package com.dreampany.crypto.data.model

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Ignore
import androidx.room.Index
import com.dreampany.framework.data.model.Base
import com.dreampany.framework.misc.constant.Constant
import com.dreampany.framework.misc.util.Util
import com.dreampany.crypto.misc.constants.AppConstants
import com.google.common.base.Objects
import kotlinx.android.parcel.Parcelize

/**
 * Created by roman on 29/2/20
 * Copyright (c) 2020 bjit. All rights reserved.
 * hawladar.roman@bjitgroup.com
 * Last modified $file.lastModified
 */
@Parcelize
//@IgnoreExtraProperties
@Entity(
    indices = [Index(
        value = [Constant.Keys.ID],
        unique = true
    )],
    primaryKeys = [Constant.Keys.ID]
)
data class Exchange(
    override var time: Long = Constant.Default.LONG,
    override var id: String = Constant.Default.STRING,
    var market: String = Constant.Default.STRING,
    @ColumnInfo(name = AppConstants.Keys.Exchange.FROM_SYMBOL)
    private var fromSymbol: String = Constant.Default.STRING,
    @ColumnInfo(name = AppConstants.Keys.Exchange.TO_SYMBOL)
    private var toSymbol: String = Constant.Default.STRING,
    var price: Double = Constant.Default.DOUBLE,
    @ColumnInfo(name = AppConstants.Keys.Exchange.VOLUME_24H)
    private var volume24h: Double = Constant.Default.DOUBLE,
    @ColumnInfo(name = AppConstants.Keys.Exchange.CHANGE_24H)
    private var change24h: Double = Constant.Default.DOUBLE,
    @ColumnInfo(name = AppConstants.Keys.Exchange.CHANGE_PCT_24H)
    private var changePct24h: Double = Constant.Default.DOUBLE
) : Base() {

    @Ignore
    constructor() : this(time = Util.currentMillis()) {

    }

    constructor(id: String) : this(time = Util.currentMillis(), id = id) {

    }

    override fun hashCode(): Int = Objects.hashCode(id)

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val item = other as Exchange
        return Objects.equal(this.id, item.id)
    }

    override fun toString(): String = "Exchange ($id) == $id"

    //@PropertyName(AppConstants.Keys.Exchange.FROM_SYMBOL)
    fun setFromSymbol(fromSymbol: String) {
        this.fromSymbol = fromSymbol
    }

    //@PropertyName(AppConstants.Keys.Exchange.FROM_SYMBOL)
    fun getFromSymbol(): String {
        return fromSymbol
    }

    //@PropertyName(AppConstants.Keys.Exchange.TO_SYMBOL)
    fun setToSymbol(toSymbol: String) {
        this.toSymbol = toSymbol
    }

    //@PropertyName(AppConstants.Keys.Exchange.TO_SYMBOL)
    fun getToSymbol(): String {
        return toSymbol
    }

    //@PropertyName(AppConstants.Keys.Exchange.VOLUME_24H)
    fun setVolume24h(volume24h: Double) {
        this.volume24h = volume24h
    }

    //@PropertyName(AppConstants.Keys.Exchange.VOLUME_24H)
    fun getVolume24h(): Double {
        return volume24h
    }

    //@PropertyName(AppConstants.Keys.Exchange.CHANGE_24H)
    fun setChange24h(change24h: Double) {
        this.change24h = change24h
    }

    //@PropertyName(AppConstants.Keys.Exchange.CHANGE_24H)
    fun getChange24h(): Double {
        return change24h
    }

    //@PropertyName(AppConstants.Keys.Exchange.CHANGE_PCT_24H)
    fun setChangePct24h(changePct24h: Double) {
        this.changePct24h = changePct24h
    }

    //@PropertyName(AppConstants.Keys.Exchange.CHANGE_PCT_24H)
    fun getChangePct24h(): Double {
        return changePct24h
    }
}